File: backend/src/main/java/com/mulesoft/arc/arcdatastore/backend/models/ArcAnalyticsDailyItemResult.java
package com.mulesoft.arc.arcdatastore.backend.models;

import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

/**
 * This object is returned in weekly and monthly analytics data.
 * It is a representation of number of users or sessions for each day for the given time period.
 */

public class ArcAnalyticsDailyItemResult implements Comparable<ArcAnalyticsDailyItemResult> {
    /**
     * A date for the value. Represented as yyyy-MM-dd
     */
    public String day;
    /**
     * Number of session or users for given day.
     */
    public int value;

    @Override
    public int compareTo(ArcAnalyticsDailyItemResult other) {
        DateFormat df = new SimpleDateFormat("yyyy-MM-dd", Locale.UK);
        Date thisDate;
        Date otherDate;
        try {
            thisDate = df.parse(this.day);
        } catch (Exception e) {
            thisDate = null;
        }
        try {
            otherDate = df.parse(other.day);
        } catch (Exception e) {
            otherDate = null;
        }
        if (thisDate == null && otherDate == null) {
            return 0;
        }
        if (thisDate == null) {
            return -1;
        }
        if (otherDate == null) {
            return 1;
        }
        return thisDate.compareTo(otherDate);
    }
}

File: models/arc-analytics-daily-item-result.js
'use strict';

/**
 * This object is returned in weekly and monthly analytics data.
 * It is a representation of number of users or sessions for each day
 * for the given time period.
 */
class ArcAnalyticsDailyItemResult {
  constructor(day, value) {
    /**
     * A date for the value. Represented as yyyy-MM-dd
     *
     * @type {String}
     */
    this.day = day;
    /**
     * Number of session or users for given day.
     *
     * @type {Number}
     */
    this.value = value;
  }

  sort(a, b) {
    const thisDate = Date.parse(a.day);
    const otherDate = Date.parse(b.day);
    if (!thisDate && !otherDate) {
      return 0;
    }
    if (!thisDate) {
      return -1;
    }
    if (!otherDate) {
      return 1;
    }
    if (thisDate > otherDate) {
      return 1;
    }
    if (thisDate < otherDate) {
      return -1;
    }
    return 0;
  }
}

module.exports.ArcAnalyticsDailyItemResult = ArcAnalyticsDailyItemResult;
